core/src/main/java/org/apache/iceberg/rest/RESTFileScanTaskParser.java

```diff
@@ -0,0 +1,109 @@
+import org.apache.iceberg.DeleteFile;
+import org.apache.iceberg.FileScanTask;
+import org.apache.iceberg.expressions.Expression;
+import org.apache.iceberg.expressions.ExpressionParser;
+import org.apache.iceberg.expressions.Expressions;
+import org.apache.iceberg.expressions.ResidualEvaluator;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+
+public class RESTFileScanTaskParser {
+  private static final String DATA_FILE = "data-file";
+  private static final String DELETE_FILE_REFERENCES = "delete-file-references";
+  private static final String RESIDUAL = "residual-filter";
+
+  private RESTFileScanTaskParser() {}
+
+  public static void toJson(
+      FileScanTask fileScanTask, List<DeleteFile> deleteFiles, JsonGenerator generator)
+      throws IOException {
+    Preconditions.checkArgument(fileScanTask != null, "Invalid file scan task: null");
+    Preconditions.checkArgument(generator != null, "Invalid JSON generator: null");
+
+    generator.writeStartObject();
+    generator.writeFieldName(DATA_FILE);
+    RESTContentFileParser.toJson(fileScanTask.file(), generator);
+
+    // TODO revisit this logic
+    if (deleteFiles != null) {
+      generator.writeArrayFieldStart(DELETE_FILE_REFERENCES);
+      for (int delIndex = 0; delIndex < deleteFiles.size(); delIndex++) {
+        generator.writeNumber(delIndex);
+      }
+      generator.writeEndArray();
+    }
+    if (fileScanTask.residual() != null) {
+      generator.writeFieldName(RESIDUAL);
+      ExpressionParser.toJson(fileScanTask.residual(), generator);
+    }
+    generator.writeEndObject();
+  }
+
+  public static FileScanTask fromJson(JsonNode jsonNode, List<DeleteFile> allDeleteFiles) {
+    Preconditions.checkArgument(jsonNode != null, "Invalid JSON node for file scan task: null");
+    Preconditions.checkArgument(
+        jsonNode.isObject(), "Invalid JSON node for file scan task: non-object (%s)", jsonNode);
+
+    DataFile dataFile = (DataFile) RESTContentFileParser.fromJson(jsonNode.get(DATA_FILE));
+
+    DeleteFile[] matchedDeleteFiles = null;
+    List<Integer> deleteFileReferences = null;
+    if (jsonNode.has(DELETE_FILE_REFERENCES)) {
+      ImmutableList.Builder deleteFileReferencesBuilder = ImmutableList.builder();
+      JsonNode deletesArray = jsonNode.get(DELETE_FILE_REFERENCES);
+      for (JsonNode deleteRef : deletesArray) {
+        deleteFileReferencesBuilder.add(deleteRef);
+      }
+      deleteFileReferences = deleteFileReferencesBuilder.build();
+    }
+
+    if (deleteFileReferences != null) {
+      ImmutableList.Builder matchedDeleteFilesBuilder = ImmutableList.builder();
+      for (Integer deleteFileIdx : deleteFileReferences) {
+        matchedDeleteFilesBuilder.add(allDeleteFiles.get(deleteFileIdx));
+      }
+      matchedDeleteFiles = (DeleteFile[]) matchedDeleteFilesBuilder.build().stream().toArray();
+    }
```

Comment:
I think the logic here for resolving the delete files for the task is incorrect and can also just be simplified to the following:



```
    Set<Integer> deleteFileReferences = Sets.newHashSet();
    if (jsonNode.has(DELETE_FILE_REFERENCES)) {
      deleteFileReferences.addAll(JsonUtil.getIntegerList(DELETE_FILE_REFERENCES, jsonNode));
    }

    List<DeleteFile> deleteFilesForTask = Lists.newArrayList();
    for (int i = 0; i < deleteFiles.size(); i++) {
      if (deleteFileReferences.contains(i)) {
        deleteFilesForTask.add(deleteFiles.get(i));
      }
    }
```

Reply:
Thanks @amogh-jahagirdar for pointing this and the concrete suggestion. 
@singhpk234 will look into adding tests for the parsers, was first just trying to get most of `src` implementation out so others can comment.